### DISCUSSION TITLE
Fixing VisualStudio performance problems by storing generated sources

### DISCUSSION THREAD

**Discussion by pziezio** [microsoft/CsWin32#359] [Show and tell]
I'd like to share my solution to VisualStudio performance problems when using CsWin32.
I'm using VS2019 Preview but I don't see that generated sources are cached in any way and IntelliSense is just not usable.
I prepared the list of the functions I use in the `NativeMethods.txt` file and I practically never add new ones anymore.

So I set `"emitSingleFile": true` in configuration file and stored generated sources in new file `NativeMethods.cs`.

My `*.csproj` file looks like this:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net5.0-windows10.0.19041.0</TargetFramework>
    <RootNamespace>CsWin32Test</RootNamespace>
    <RuntimeIdentifiers>win10-x64</RuntimeIdentifiers>
    <Platforms>x64</Platforms>
    <UseCsWin32>false</UseCsWin32>
  </PropertyGroup>
  <ItemGroup Condition="'$(UseCsWin32)'=='true'">
    <Compile Remove="NativeMethods.cs" />
    <PackageReference Include="Microsoft.Windows.CsWin32" Version="0.1.529-beta-ge5d36b007f">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
  </ItemGroup>
</Project>
```

I can set custom property `UseCsWin32` to `true` if I want to generate new imports or update packages.
Of course, it contradicts the idea of automatic code generation but it just works for me.

Idea: maybe it would be possible to add new configuration option like `"cacheSources": true` that would store generated sources on disk. That would solve all of the performance problems once and for all.

**Comment by jnm2** [comment]
See https://github.com/microsoft/CsWin32/pull/355 also.

**Comment by AArnott** [comment]
We made some significant perf improvements a short while back. What version are you using when you experience unacceptable perf degradation, @pziezio?

**Reply by pziezio** [reply]
I'm always using the most current version and updating as soon as possible. I've tested version compiled from the sources too.
Still using VS2019 though. I will try incremental code generation with VS2022 when it'll be ready.

**Reply by AArnott** [reply]
Can you collect an ETL trace of the problem and share with me? You can use Report a Problem in VS and follow the steps for recording a perf problem there, then send me a link to the ticket you file.
